src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.OptionTracking.cs

```diff
@@ -0,0 +1,198 @@
+
+                // Ignore it during the processing of int-value options.
+                trackedOptions &= ~lingerMask;
+            }
+
+            // For DualMode, we use the value stored in the handle rather than querying the socket itself,
+            // as on Unix stacks binding a dual-mode socket to an IPv6 address may cause IPV6_V6ONLY to revert to true.
+            int ipv6OnlyMask = GetMask(TrackableSocketOptions.IPV6_V6ONLY);
+            if ((trackedOptions & ipv6OnlyMask) == ipv6OnlyMask)
+            {
+                values[(int)TrackableSocketOptions.IPV6_V6ONLY - 1] = DualMode ? 0 : 1;
+                trackedOptions &= ~ipv6OnlyMask;
+            }
+
+            for (int i = 0; i < values.Length; i++)
+            {
+                int mask = 1 << i;
+                if ((trackedOptions & mask) == mask)
+                {
+                    TrackableSocketOptions tracked = (TrackableSocketOptions)(i + 1);
+                    (SocketOptionName name, SocketOptionLevel level) = ToSocketOptions(tracked);
+                    SocketError errorCode = SocketPal.GetSockOpt(this, level, name, out values[i]);
+                    if (NetEventSource.Log.IsEnabled() && errorCode != SocketError.Success) NetEventSource.Info(this, $"GetSockOpt({level},{name}) returned errorCode:{errorCode}");
+                }
+            }
+        }
+
+        internal void SetTrackedSocketOptions(ReadOnlySpan<int> values, LingerOption? lingerOption)
+        {
+            int lingerMask = GetMask(TrackableSocketOptions.SO_LINGER);
+            if (lingerOption is not null)
+            {
+                Debug.Assert((_trackedOptions & lingerMask) == lingerMask);
+                SocketError errorCode = SocketPal.SetLingerOption(this, lingerOption);
+                if (NetEventSource.Log.IsEnabled() && errorCode != SocketError.Success) NetEventSource.Info(this, $"SetLingerOption returned errorCode:{errorCode}");
+            }
+
+            int trackedOptions = _trackedOptions & ~lingerMask;
+
+            for (int i = 0; i < values.Length; i++)
+            {
+                int mask = 1 << i;
+                if ((trackedOptions & mask) == mask)
+                {
+                    TrackableSocketOptions tracked = (TrackableSocketOptions)(i + 1);
+                    (SocketOptionName name, SocketOptionLevel level) = ToSocketOptions(tracked);
+                    SocketError errorCode = SocketPal.SetSockOpt(this, level, name, values[i]);
+                    if (NetEventSource.Log.IsEnabled() && errorCode != SocketError.Success) NetEventSource.Info(this, $"GetSockOpt({level},{name}) returned errorCode:{errorCode}");
+                }
+            }
+        }
+
+        private static int GetMask(TrackableSocketOptions tracked) => 1 << ((int)tracked - 1);
+
+        // Relevant option names and values taken from Windows headers.
```

Comment:
While a source generator would be an overkill, the code below this line is generated by https://gist.github.com/antonfirsov/2cbfc37e665ad840ed7734994948c29a. I wonder if it's ok to link the gist from the code or add the codegen to the repo in some other way.

Reply:
IMO it'd be fine, we [did it already](https://grep.app/search?q=gist.github.com/&filter[repo][0]=dotnet/runtime) in a couple places